tests/unittests/test_runs/test_fallback_datasource.py

```diff
@@ -0,0 +1,240 @@
+        if path in os_files:
+            return (200, headers, os_files.get(path))
+        return (404, headers, '')
+
+    def get_request_callback(method, uri, headers):
+        uri = urlparse(uri)
+        path = uri.path.lstrip("/").split("/")
+        if path[0] == 'openstack':
+            return match_os_uri(uri, headers)
+        return match_ec2_url(uri, headers)
+
+    hp.register_uri(hp.GET, re.compile(r'http://169.254.169.254/.*'),
+                    body=get_request_callback)
+
+
+class TestFallbackDatasource(tests_helpers.FilesystemMockingTestCase,
+                             tests_helpers.HttprettyTestCase):
+    VERSION = 'latest'
+
+    def setUp(self):
+        super(TestFallbackDatasource, self).setUp()
+        self.tmp = self.tmp_dir()
+        self.paths = helpers.Paths(
+            {'cloud_dir': self.tmp, 'run_dir': self.tmp})
+        self.ds = dscd.DataSourceConfigDrive
+
+    def cloud_init(self, initer):
+        initer.read_cfg()
+        initer.initialize()
+        initer.fetch()
+        iid = initer.instancify()
+        initer.update()
+        return iid
+
+    def remove_seed_dir(self, seed_dir):
+        self.patched_funcs.close()
+        util.del_dir(seed_dir)
+        util.del_file(os.path.join(self.tmp, "run/cloud-init/.instance-id"))
+        util.del_file(os.path.join(self.tmp, ".instance-id"))
+
+    def test_invalid_config_drive(self):
+        seed_dir = os.path.join(self.paths.seed_dir, "config_drive")
+        populate_dir(seed_dir, CFG_DRIVE_FILES_V2)
+
+        sys_cfg = {
+            'datasource_list': ['ConfigDrive'],
+            'datasource': {'ConfigDrive': {'dsmode': 'local'}}
+        }
+
+        cloud_cfg = safeyaml.dumps(sys_cfg)
+        util.ensure_dir(os.path.join(self.tmp, 'etc', 'cloud'))
+        util.write_file(os.path.join(self.tmp, 'etc',
+                                     'cloud', 'cloud.cfg'), cloud_cfg)
+        dsrc = self.ds(sys_cfg=sys_cfg, distro=None, paths=self.paths)
+        ret = dsrc.get_data()
+        self.assertEqual(
+            dsrc.subplatform, 'seed-dir (%s)' % seed_dir)
+        self.assertTrue(ret)
+        # simulate first boot with config drive
+        self.reRoot(self.tmp)
+        initer = stages.Init()
+        initer.datasource = dsrc
+        iid_with_dsrc = self.cloud_init(initer)
+        # simulate subsequent boot without config drive
+        self.remove_seed_dir(seed_dir)
+        self.reRoot(self.tmp)
+        initer = stages.Init(ds_deps=['FILESYSTEM'])
+        iid_without_dsrc = self.cloud_init(initer)
+        self.assertEqual(iid_with_dsrc, iid_without_dsrc)
+
+    def test_invalid_openstack_uri(self):
+        _register_uris(self.VERSION, {}, {}, OS_FILES)
+        sys_cfg = {
+            'datasource_list': ['OpenStack'],
+            'datasource': {'OpenStack': {'dsmode': 'net'}}
+        }
+        cloud_cfg = safeyaml.dumps(sys_cfg)
+        util.ensure_dir(os.path.join(self.tmp, 'etc', 'cloud'))
```

Comment:
You can drop this ensure_dir, write_file does the equivalent of: mkdir -p $(dirname filename)

Reply:
Done